### PR TITLE
Updated playbook dependencies optional branch

### PR DESCRIPTION
**Pull request by yaakovi**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/26065

## Description
when a task is marked with `skipunavailable`, mark rest playbook branch as an optional dependency.

### CODE REVIEW

File: demisto_sdk/commands/common/constants.py [hunk 1]
```diff
@@ -841,7 +841,7 @@ class PB_Status:
 BASE_PACK = "Base"
 NON_SUPPORTED_PACK = "NonSupported"
 DEPRECATED_CONTENT_PACK = "DeprecatedContent"
-IGNORED_DEPENDENCY_CALCULATION = {BASE_PACK, NON_SUPPORTED_PACK, DEPRECATED_CONTENT_PACK}
+IGNORED_DEPENDENCY_CALCULATION = {BASE_PACK, NON_SUPPORTED_PACK, }
```
Comment (yaakovi) [inline demisto_sdk/commands/common/constants.py:844]: Temporary for local testing - will be reverted

File: demisto_sdk/commands/common/update_id_set.py [hunk 2]
```diff
@@ -136,6 +137,60 @@ def get_changed_files(files_string):
     return added_files_list, modified_files_list, added_script_list, modified_script_list
 
 
+def build_dependency_graph(playbook_data):
```
Comment (IkaDemisto) [inline demisto_sdk/commands/common/update_id_set.py:140]: Function name can be confusing. I think tasks graph is more suitable here

File: demisto_sdk/commands/common/update_id_set.py [hunk 3]
```diff
@@ -136,6 +137,60 @@ def get_changed_files(files_string):
...
+        for leaf in leaf_nodes:
+            leaf_task = tasks.get(leaf)
+            leaf_mandatory = graph.nodes[leaf]['mandatory']
+
+            # In this case the playbook is invalid, one of the "nexttasks" fields contain invalid task id.
+            if not leaf_task:
+                print_error(f'{playbook_data.get("id")}: No such task {leaf} in playbook')
+                continue
+                # raise ValueError(f'No such task {leaf} in playbook {playbook_data.get("id")}')
+            leaf_next_tasks = sum(leaf_task.get('nexttasks', {}).values(), [])
```
Comment (IkaDemisto) [inline demisto_sdk/commands/common/update_id_set.py:170]: Not sure why you need to use sum on `nexttasks`

File: demisto_sdk/commands/common/update_id_set.py [hunk 4]
```diff
@@ -136,6 +137,60 @@ def get_changed_files(files_string):
...
+                    # raise ValueError(f'No such task {leaf} in playbook {playbook_data.get("id")}')
+
+                # If task can't be skipped and predecessor task is mandatory - set as mandatory.
+                mandatory = leaf_mandatory and not task.get('skipunavailable', False)
+                if task_id not in graph.nodes():
+                    graph.add_node(task_id, mandatory=mandatory)
+                else:
+                    # If task already in graph, update mandatory field.
+                    # If one of the paths to the task is mandatory - set as mandatory.
+                    graph.nodes[task_id]['mandatory'] = graph.nodes[task_id]['mandatory'] or mandatory
```
Comment (IkaDemisto) [inline demisto_sdk/commands/common/update_id_set.py:186]: Should not it be just:
```
graph.nodes[task_id]['mandatory'] = mandatory
```
?

File: demisto_sdk/commands/common/update_id_set.py (hunk 3 above)
Comment (yaakovi) [reply demisto_sdk/commands/common/update_id_set.py:170]: in conditional tasks, there might be several lists:
image

File: demisto_sdk/commands/common/update_id_set.py (hunk 4 above)
Comment (yaakovi) [reply demisto_sdk/commands/common/update_id_set.py:186]: just to be on the safe side,
if I first recognized it as mandatory and then I found another task that is optional, I don't want to override the mandatory mark. this way I can maintain the `mandatory: true` once I found it.

File: demisto_sdk/commands/common/update_id_set.py [hunk 5]
```diff
@@ -137,6 +138,60 @@ def get_changed_files(files_string):
...
+
+        for leaf in leaf_nodes:
+            leaf_task = tasks.get(leaf)
+            leaf_mandatory = graph.nodes[leaf]['mandatory']
+
+            # In this case the playbook is invalid, one of the "nexttasks" fields contain invalid task id.
+            if not leaf_task:
+                print_error(f'{playbook_data.get("id")}: No such task {leaf} in playbook')
+                continue
+                # raise ValueError(f'No such task {leaf} in playbook {playbook_data.get("id")}')
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:170]: Remove this comment if not needed

File: demisto_sdk/commands/common/update_id_set.py [hunk 6]
```diff
@@ -137,6 +138,60 @@ def get_changed_files(files_string):
...
+            leaf_next_tasks = sum(leaf_task.get('nexttasks', {}).values(), [])
+
+            for task_id in leaf_next_tasks:
+                task = tasks.get(task_id)
+                if not task:
+                    print_error(f'{playbook_data.get("id")}: No such task {leaf} in playbook')
+                    continue
+                    # raise ValueError(f'No such task {leaf} in playbook {playbook_data.get("id")}')
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:178]: Same

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 7]
```diff
@@ -250,22 +270,26 @@ def _label_as_optional(pack_ids):
...
         Returns:
             set: dependencies data that includes pack id and whether is mandatory or not.
 
         """
         dependencies_packs = set()
+        verbose_file.write('\n### Scripts')
 
         for script_mapping in pack_scripts:
             script = next(iter(script_mapping.values()))
+            script_dep = set()
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:291]:
```suggestion
            script_dependencies = set()
```

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 8]
```diff
@@ -275,99 +299,153 @@ def _collect_scripts_dependencies(pack_scripts, id_set):
 
                 if pack_name:  # found script dependency implementing pack name
                     pack_dependencies_data = PackDependencies._label_as_mandatory(pack_name)
-                    dependencies_packs.update(pack_dependencies_data)  # set found script as mandatory
+                    script_dep.update(pack_dependencies_data)  # set found script as mandatory
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:302]:
```suggestion
                    script_dependencies.update(pack_dependencies_data)  # set found script as mandatory
```

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 9]
```diff
@@ -275,99 +299,153 @@ def _collect_scripts_dependencies(pack_scripts, id_set):
...
                     continue  # found dependency in script section, skipping to next depends on element
 
                 # try to search dependency by integration integration
                 pack_names = PackDependencies._search_packs_by_integration_command(command, id_set)
 
                 if pack_names:  # found integration dependency implementing pack name
                     pack_dependencies_data = PackDependencies._detect_generic_commands_dependencies(pack_names)
-                    dependencies_packs.update(pack_dependencies_data)
+                    script_dep.update(pack_dependencies_data)
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:310]:
```suggestion
                    script_dependencies.update(pack_dependencies_data)
```

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 10]
```diff
@@ -275,99 +299,153 @@ def _collect_scripts_dependencies(pack_scripts, id_set):
...
+
+            verbose_file.write(
+                f'{os.path.basename(script.get("file_path", ""))} depends on: '
+                f'{script_dep}  '
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:314]:
```suggestion
                f'{script_dependencies}  '
```

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 11]
```diff
@@ -275,99 +299,153 @@ def _collect_scripts_dependencies(pack_scripts, id_set):
...
+            )
+            dependencies_packs.update(script_dep)
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:316]:
```suggestion
            dependencies_packs.update(script_dependencies)
```

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 12]
```diff
@@ -385,36 +463,45 @@ def _collect_layouts_dependencies(pack_layouts, id_set):
...
 
-            related_incident_types = incident_field_data.get('incident_types', [])
-            packs_found_from_incident_types = PackDependencies._search_packs_by_items_names(
-                related_incident_types, id_set['IncidentTypes'])
-
-            if packs_found_from_incident_types:
-                pack_dependencies_data = PackDependencies. \
-                    _label_as_mandatory(packs_found_from_incident_types)
-                dependencies_packs.update(pack_dependencies_data)
+            # related_incident_types = incident_field_data.get('incident_types', [])
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:497]: Why did you comment this section?

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 13]
```diff
@@ -717,22 +891,24 @@ def build_dependency_graph(pack_id, id_set):
...
     @staticmethod
-    def find_dependencies(pack_name, id_set_path=None):
+    def find_dependencies(pack_name, id_set_path=None, debug_file_path=None):
         """
         Main function for dependencies search and pack metadata update.
 
         Args:
             pack_name (str): pack id, currently pack folder name is in use.
             id_set_path (str): id set json.
+            debug_file_path (str): path to dependency explanations file.
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:901]: Doesn't mypy return error here?
id_set_path, debug_file_path are of type `str` while their default value is None.
Consider changing their default to `''`

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 14]
```diff
@@ -717,22 +891,24 @@ def build_dependency_graph(pack_id, id_set):
...
         """
         if not id_set_path:
-            id_set = IDSetCreator(output=None, print_logs=False).create_id_set()
+            id_set = IDSetCreator(print_logs=False).create_id_set()
         else:
             with open(id_set_path, 'r') as id_set_file:
                 id_set = json.load(id_set_file)
-
-        dependency_graph = PackDependencies.build_dependency_graph(pack_id=pack_name, id_set=id_set)
+        with VerboseFile(debug_file_path) as verbose_file:
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:909]: What happens if the `debug_file_path` does not exist?

File: demisto_sdk/commands/find_dependencies/find_dependencies.py (hunk 12 above)
Comment (IkaDemisto) [reply demisto_sdk/commands/find_dependencies/find_dependencies.py:497]: +1

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 15]
```diff
@@ -275,99 +299,153 @@ def _collect_scripts_dependencies(pack_scripts, id_set):
...
             pack_playbooks (list): collection of pack playbooks data.
             id_set (dict): id set json.
+            verbose_file (VerboseFile): path to dependency explanations file.
 
         Returns:
             set: dependencies data that includes pack id and whether is mandatory or not.
 
         """
         dependencies_packs = set()
+        verbose_file.write('\n### Playbooks')
```
Comment (IkaDemisto) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:367]: Good idea. I think in later step the log should be extended and contain the reason and additional data for found dependency.

File: demisto_sdk/commands/find_dependencies/find_dependencies.py (hunk 12 above)
Comment (yaakovi) [reply demisto_sdk/commands/find_dependencies/find_dependencies.py:497]: after speaking with @michalgold we agreed that this dependency is not correct.
fields should be based on types as they can exist without it.

File: demisto_sdk/commands/find_dependencies/find_dependencies.py (hunk 14 above)
Comment (yaakovi) [reply demisto_sdk/commands/find_dependencies/find_dependencies.py:909]: the object ignores all the prints, see the implementation at the top of the file

File: demisto_sdk/commands/find_dependencies/find_dependencies.py (hunk 15 above)
Comment (yaakovi) [reply demisto_sdk/commands/find_dependencies/find_dependencies.py:367]: what do you mean later step?
this is output example:

image

Comment (IkaDemisto) [reply demisto_sdk/commands/find_dependencies/find_dependencies.py:367]: I mean that if layout from pack A uses specific field from pack B then better to log it somewhere. But it is not urgent.
For example: https://panw-global.slack.com/archives/G011E63JXPB/p1594907099182700

**Pull request merged**